lib/url-pattern/url-match-pattern.js

```diff
@@ -0,0 +1,279 @@
+var _ = require('../util').lodash,
+    Property = require('../collection/property').Property,
+    Url = require('../collection/url').Url,
+
+    MATCH_ALL = '*',
+    PREFIX_DELIMITER = '^',
+    POSTFIX_DELIMITER = '$',
+    MATCH_ALL_URLS = '<all_urls>',
+    ALLOWED_PROTOCOLS = ['http', 'https', 'file', 'ftp'],
+    ALLOWED_PROTOCOLS_REGEX = ALLOWED_PROTOCOLS.join('|'),
+
+    regexes = {
+        escapeMatcher: /[.+^${}()|[\]\\]/g,
+        escapeMatchReplacement: '\\$&',
+        questionmarkMatcher: /\?/g,
+        questionmarkReplacment: '.',
+        starMatcher: '*',
+        starReplacement: '.*',
+        patternSplit: '^(' + ALLOWED_PROTOCOLS_REGEX + '|\\*)*://(\\*|\\*\\.[^*/]+|[^*/]+|)(/.*)$'
+    },
+
+    UrlMatchPattern;
+
+/**
+ * UrlMatchPattern allows to create rules to define Urls to match for.
+ * It is based on Google's Match Pattern - https://developer.chrome.com/extensions/match_patterns
+ *
+ * @example <caption>An example UrlMatchPattern</caption>
+ * var matchPattern = new UrlMatchPattern('https://*.google.com/*');
+*/
+_.inherit((
+  UrlMatchPattern = function UrlMatchPattern (pattern) {
+      UrlMatchPattern.super_.apply(this, { pattern: pattern });
+
+      this.pattern = pattern && _.isString(pattern) ? pattern : MATCH_ALL_URLS;
+  }), Property);
+
+_.assign(UrlMatchPattern.prototype, /** @lends UrlMatchPattern.prototype */ {
+    /**
+     * Used to generate the match regex object from the match string we have.
+     * @private
+     * @returns Match regex object
+     */
+    createMatchPattern: function() {
+        var matchPattern = this.pattern,
+
+            // Check the match pattern of sanity and split it into protocol, host and path
+            match = matchPattern.match(regexes.patternSplit);
+
+        if (!match) {
+            // This ensures it is a invalid match pattern
+            return undefined;
+        }
+
+        return ({
+            protocol: match[1],
+            host: match[2],
+            path: this.globPatternToRegexp(match[3])
+        });
+    },
+
+    /**
+     * Converts a given glob pattern into a regular expression
+     * @private
+     * @param {String} pattern Glob pattern string
+     * @return {RegExp=}
+     */
+    globPatternToRegexp: function(pattern) {
+        // Escape everything except ? and *.
+        pattern = pattern.replace(regexes.escapeMatcher, regexes.escapeMatchReplacement);
+        pattern = pattern.replace(regexes.questionmarkMatcher, regexes.questionmarkReplacment);
+        pattern = pattern.replace(regexes.starMatcher, regexes.starReplacement);
+        return new RegExp(PREFIX_DELIMITER + pattern + POSTFIX_DELIMITER);
```

Comment:
In this case (regex with recursive match) a static function with replacements is slower than individual replacements.

https://jsperf.com/string-replace-vs-replace-function